program/src/processor/mod.rs

```diff
@@ -144,3 +161,79 @@ macro_rules! require {
         require!( $constraint, $error, format!($message, $($args)+) );
     };
 }
+
+/// Unpacks an initialized account from the given data and
+/// returns a mutable reference to it.
+#[macro_export]
+macro_rules! unpack_initialized_mut {
+    ( $data:expr, $type:ty, $name:literal ) => {{
+        let account = bytemuck::try_from_bytes_mut::<$type>($data)
+            .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+        require!(
+            account.is_initialized(),
+            ProgramError::UninitializedAccount,
+            $name,
+        );
+
+        account
+    }};
+}
+
+/// Unpacks the delegation information from either a `SolStakerStake` and `ValidatorStake`
+/// accounts.
+///
+/// This function will validate that the account data is initialized and deriavation matches
+/// the expected PDA derivation.
+#[inline]
+pub fn unpack_delegation_mut(
+    stake_data: &mut [u8],
+    stake: &Pubkey,
+    config: &Pubkey,
+    program_id: &Pubkey,
+) -> Result<Delegation, ProgramError> {
```

Comment:
The changes are done in place since the `Delegation` is moved out of the stake struct, so the caller gets "full" ownership of it. We probably can't return a reference since the stake object will be out of the scope after the function returns.

Reply:
I don't think this is the case. I ran a little test on the sol stake view program, check it out: https://github.com/paladin-bladesmith/sol-stake-view-program/pull/4

Reply:
You are absolutely right! `Pod` types must implement `Copy`, so a copy is being returned. 🤦‍♂️

Reply:
We can't return a reference either.